src/main/scala/zio/s3/Live.scala

```diff
@@ -146,81 +146,97 @@ final class Live(unsafeClient: S3AsyncClient) extends S3 {
-      _        <-
+  def multipartUploadSink(bucketName: String, key: String, options: MultipartUploadOptions)(
+    parallelism: Int
+  ): ZSink[Any, S3Exception, Byte, Byte, Unit] = {
+
+    val startUpload = ZSink.fromZIO(
+      execute(
+        _.createMultipartUpload {
+          val builder = CreateMultipartUploadRequest
+            .builder()
+            .bucket(bucketName)
+            .key(key)
+            .metadata(options.uploadOptions.metadata.asJava)
+            .acl(options.uploadOptions.cannedAcl)
+          options.uploadOptions.contentType
+            .fold(builder)(builder.contentType)
+            .build()
+        }
+      )
+        .map(_.uploadId())
+    )
+
+    def completeUpload(uploadId: String, parts: Chunk[CompletedPart]) =
+      ZSink.fromZIO(
+        execute(
+          _.completeMultipartUpload(
+            CompleteMultipartUploadRequest
+              .builder()
+              .bucket(bucketName)
+              .key(key)
+              .multipartUpload(CompletedMultipartUpload.builder().parts(parts.asJavaCollection).build())
+              .uploadId(uploadId)
+              .build()
+          )
+        )
+      )
+
+    def uploadParts(uploadId: String): ZSink[Any, S3Exception, Byte, Nothing, Chunk[CompletedPart]] = {
+      def go(
+        hasFirst: Boolean,
+        partNumber: Int
+      ): ZChannel[Any, S3Exception, Chunk[Byte], Any, S3Exception, (Chunk[Byte], Int), Unit] =
+        ZChannel.readWith(
+          (in: Chunk[Byte]) => ZChannel.write(in -> partNumber) *> go(hasFirst = true, partNumber + 1),
+          (err: S3Exception) => ZChannel.fail(err),
+          (_: Any) =>
+            if (hasFirst) ZChannel.unit
+            else ZChannel.write(Chunk.empty -> partNumber) *> ZChannel.unit
+        )
+
+      def writer(
+        uploadId: String
+      ): ZChannel[Any, Nothing, Chunk[Byte], Any, S3Exception, Chunk[CompletedPart], Unit] =
+        go(false, 1)
+          .mapOutZIOPar(parallelism) {
+            case (chunk, partNumber) =>
+              execute(
+                _.uploadPart(
+                  UploadPartRequest
+                    .builder()
+                    .bucket(bucketName)
+                    .key(key)
+                    .partNumber(partNumber + 1)
+                    .uploadId(uploadId)
+                    .contentLength(chunk.length.toLong)
+                    .build(),
+                  AsyncRequestBody.fromBytes(chunk.toArray)
+                )
+              ).map(r => Chunk.single(CompletedPart.builder().partNumber(partNumber.toInt + 1).eTag(r.eTag()).build()))
+          }
+
+      ZSink.fromChannel(writer(uploadId)).collectLeftover.map(_._2)
+    }
+
+    val writeThrough = ZSink.fromChannel(ZChannel.identity[S3Exception, Chunk[Byte], Any])
+
+    val uploader = for {
+      uploadId <- startUpload
+      parts    <- uploadParts(uploadId) zipParLeft writeThrough
```

Comment:
I'm not sure if there is a better way to preserve the write through here (or if it is even necessary), since I know that `zipPar` is a fairly heavy operation on a `Sink`